File: src/main/java/com/droiddevgeeks/crashsink/AndroidDeviceMetadata.java
package com.droiddevgeeks.crashsink;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.os.Build;

/** Collects {@link DeviceMetadata} from an Android Context (off the crash path). */
public final class AndroidDeviceMetadata {

    private AndroidDeviceMetadata() {
    }

    public static DeviceMetadata collect(final Context context) {
        String appVersionName = "unknown";
        int appVersionCode = 0;
        try {
            final PackageInfo pi =
                    context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            if (pi.versionName != null) {
                appVersionName = pi.versionName;
            }
            appVersionCode = pi.versionCode;
        } catch (Throwable t) {
            // leave defaults
        }
        return new DeviceMetadata(
                Build.VERSION.RELEASE, Build.VERSION.SDK_INT,
                Build.MANUFACTURER, Build.MODEL,
                appVersionName, appVersionCode);
    }
}

File: crashsink/src/main/kotlin/com/droiddevgeeks/crashsink/AndroidDeviceMetadata.kt
package com.droiddevgeeks.crashsink

import android.content.Context
import android.os.Build

/** Collects [DeviceMetadata] from an Android Context (off the crash path). Internal. */
internal object AndroidDeviceMetadata {

    fun collect(context: Context): DeviceMetadata {
        var appVersionName = "unknown"
        var appVersionCode = 0
        try {
            val pi = context.packageManager.getPackageInfo(context.packageName, 0)
            val name = pi.versionName
            if (name != null) {
                appVersionName = name
            }
            @Suppress("DEPRECATION")
            appVersionCode = pi.versionCode
        } catch (t: Throwable) {
            // leave defaults
        }
        return DeviceMetadata(
            Build.VERSION.RELEASE, Build.VERSION.SDK_INT,
            Build.MANUFACTURER, Build.MODEL,
            appVersionName, appVersionCode
        )
    }
}
